app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java

```diff
@@ -22,110 +22,104 @@
-                Log.i("Glucosio", "Done exporting readings");
+    public void createCSVFile(@NonNull List<GlucoseReading> readings, @NonNull OutputStreamWriter osw) throws IOException {
+        // CSV Structure
+        // Date | Time | Concentration | Unit | Measured | Notes
+        Resources resources = context.getResources();
+        writeLine(osw,
+                resources.getString(R.string.dialog_add_date),
+                resources.getString(R.string.dialog_add_time),
+                resources.getString(R.string.dialog_add_concentration),
+                resources.getString(R.string.helloactivity_spinner_preferred_glucose_unit),
+                resources.getString(R.string.dialog_add_measured),
+                resources.getString(R.string.dialog_add_notes)
+        );
+
+        // Concentration | Measured | Date | Time | Notes | Unit of Measurement
+        if (Constants.Units.MG_DL.equals(userMeasurements)) {
+            for (GlucoseReading reading : readings) {
+                writeLine(osw,
+                        dateTool.convertRawDate(reading.getCreated()),
+                        dateTool.convertRawTime(reading.getCreated()),
+                        formatter.format(reading.getReading()),
+                        Constants.Units.MG_DL,
+                        valueOrEmptyString(reading.getReading_type()),
+                        valueOrEmptyString(reading.getNotes())
+                );
+
             }
-            realm.close();
-            return file == null ? null : file.getPath();
-        } catch (Exception e) {
-            realm.close();
-            e.printStackTrace();
-            return null;
+        } else {
+            for (GlucoseReading reading : readings) {
+                writeLine(osw,
+                        this.dateTool.convertRawDate(reading.getCreated()),
+                        this.dateTool.convertRawTime(reading.getCreated()),
+                        formatter.format(GlucosioConverter.glucoseToMmolL(reading.getReading())),
+                        Constants.Units.MMOL_L,
+                        valueOrEmptyString(reading.getReading_type()),
+                        valueOrEmptyString(reading.getNotes())
+                );
+            }
+
         }
+        osw.flush();
+        Log.i("Glucosio", "Done exporting readings");
+    }
+
+    @NonNull
+    private String valueOrEmptyString(@Nullable String value) {
+        return value == null ? "" : value;
     }
 
-    private void writeLine(OutputStreamWriter osw, String... values) throws IOException {
+    private void writeLine(@NonNull OutputStreamWriter osw, @NonNull String... values) throws IOException {
         for (int i = 0; i < values.length; i++) {
-            osw.append(values[i]);
+            osw.append(escapeCSV(values[i]));
             osw.append(i == values.length - 1 ? '\n' : ',');
         }
     }
+
+    private String escapeCSV(String value) {
+        if (containsSymbolsToEscape(value)) {
```

Comment:
```SYMBOLS_TO_ESCAPE``` actually contains more symbols that get checked (and lead to a return true) than the following line will actually escape.

Reply:
Can you elaborate more? I don't get the comment, sorry.

Did I make some logical error?

Reply:
I'll try :) containsSymbolsToEscape checks for a certain set of characters while not all of them get escaped then.

Reply:
Can you be more specific? What characters are not escaped?

Reply:
I'll try :)

```
 if (containsSymbolsToEscape(value)) {
    return "\"" + value.replace("\"", "\"\"") + "\"";
} 
```
escaping is done for ```"``` while the condition ```containsSymbolsToEscape(value)``` needs to be met of course. Thus I expected the condition to check for the same pattern. But it actually checks for:
```
private boolean containsSymbolsToEscape(String value) {
        for (char c : SYMBOLS_TO_ESCAPE) {
            if (value.indexOf(c) > -1) {
                return true;
            }
        }

        return false;
    }
```
while ```SYMBOLS_TO_ESCAPE``` contains:
```
private static final char[] SYMBOLS_TO_ESCAPE = new char[]{',', '"', '\n', '\r'};
```` 
So I would have expected the actual ascaping to also then escape the other "symbols to escape".

Reply:
other than that also when ```\n``` has been found the check return true and you are escaping for ```"```

Reply:
If `,` or `\n` or `\r` is detected then escaping is wrapping the whole string in the single quotes `"`. However, if the string contains single quote it should be escaped as double quote `""`. I hope it helps.

Reply:
Ah 👍 
I didn't think of that, now makes send. Thanks for the explanation and your patience 😃

Reply:
No issues, I also didn't know it by heart and should, of course, mention it in PR description.